src/Tools/dotnet-monitor/Egress/Extension/ExtensionEgressProvider.cs

```diff
@@ -50,57 +47,15 @@ public ExtensionEgressProvider(IEgressPropertiesProvider propertyProvider, ILogg
                 ProviderName = providerName,
             };
 
-            /* [TODOs]
-             * 1. Add a new service to dynamically find these extension(s)
-             * 2. Remove all raw logging statements from this method and refactor into LoggingExtensions
-             * 3. Stream StdOut and StdErr async in the process so their streams don't need to end before we can return
-             * 4. Refactor WaitForExit to do an async wait
-             * 5. Add well-factored protocol for returning information from an extension
-             */
-            const string extensionProcessPath = "MyPathTo\\Extension.exe";
-            ProcessStartInfo pStart = new ProcessStartInfo()
-            {
-                FileName = extensionProcessPath,
-                RedirectStandardInput = true,
-                RedirectStandardOutput = true,
-                RedirectStandardError = true,
-                CreateNoWindow = true,
-                UseShellExecute = false,
-            };
-            Process p = new Process()
-            {
-                StartInfo = pStart,
-            };
-
-            Logger?.LogInformation("Starting process...");
-            p.Start();
-            Logger?.LogInformation("starting stream...");
-
-            await JsonSerializer.SerializeAsync<ExtensionEgressPayload>(p.StandardInput.BaseStream, payload, options: null, token);
-            await p.StandardInput.WriteLineAsync();
-            await p.StandardInput.BaseStream.FlushAsync(token);
-
-            await action(p.StandardInput.BaseStream, token);
-            await p.StandardInput.BaseStream.FlushAsync(token);
-
-            Logger?.LogInformation("Stream done, sending close...");
-
-            p.StandardInput.Close();
-
-            Logger?.LogInformation("Waiting for exit...");
-
-            string procStdOutContent = await p.StandardOutput.ReadToEndAsync();
-            string procErrorContent = await p.StandardError.ReadToEndAsync();
-
-            await p.WaitForExitAsync(token);
-            Logger?.LogInformation($"Exited with code: {p.ExitCode}");
+            IEgressExtension ext = _extensionDiscoverer.FindExtension<IEgressExtension>(providerType);
+            EgressArtifactResult result = await ext.EgressArtifact(payload, action, token);
 
-            if (!string.IsNullOrWhiteSpace(procErrorContent))
+            if (!result.Succeeded)
```

Comment:
Based on your TODO comment, I think you're still working on the `EgressArtifact` method (so you can feel free to just resolve this if it's not relevant), but depending on how that gets implemented, should we have a null check on `result` before trying to check if it succeeded?

Reply:
You are correct, I moved this TODO to `ProgramExtension.cs` which will actually handle `EgressArtifact`.

Reply:
Based on the discussion last week I made this throw rather than returning `null`